task/blue-green-deploy/0.1/blue-green-deploy.yaml

```diff
@@ -0,0 +1,130 @@
+  steps:
+    - name: change-deployment
+      image: quay.io/openshift/origin-cli:latest
+      workingDir: $(workspaces.manifest-dir.path)
+      script: |
+        #!/usr/bin/env bash
+
+        if [[ -f $(workspaces.kubeconfig-dir.path)/kubeconfig ]]; then
+          export KUBECONFIG=$(workspaces.kubeconfig-dir.path)/kubeconfig
+        fi
+
+        getReasons() {
+          local deploy_name=$1
+          kubectl get deploy "$deploy_name" \
+            -o jsonpath='{range .status.conditions[*]}{.reason}{"\n"}{end}'
+        }
+
+        # Getting the status at particular index
+        getStatusAtIndex() {
+          local deploy_name=$1
+          local index=$2
+          kubectl get deploy "$deploy_name" \
+            -o jsonpath='{range .status.conditions['$index']}{.status}{"\n"}{end}'
+        }
+
+        # Checking for unavailable pods in case the deployment is
+        # patched with a newer image
+        checkUnavailableReplicas() {
+          local deploy_name=$1
+          local replicasUnavailable=$(kubectl get deploy "$deploy_name" -o jsonpath='{.status.unavailableReplicas}')
+          [[ "$replicasUnavailable" == "" ]] && return 0 || return 1
+        }
+
+        getMinimumReplicaStatus() {
+          local deploy_name=$1
+          local INDEX=0
+          for reason in $(getReasons $deploy_name); do
+            if [[ "$reason" == "MinimumReplicasAvailable" ]]; then
+              local ready=$(getStatusAtIndex $deploy_name ${INDEX})
+              [[ "$ready" != "True" ]] && return 1 || return 0
+            fi
+            ((INDEX++))
+          done
+          return 1
+        }
+
+        checkProgressDeadline() {
+          local deploy_name=$1
+          local INDEX=0
+          for reason in $(getReasons $deploy_name); do
+            if [[ "$reason" == "ProgressDeadlineExceeded" ]]; then
+              local status=$(getStatusAtIndex $deploy_name ${INDEX})
+              [[ "$status" != "False" ]] && return 0 || return 1
+            fi
+            ((INDEX++))
+          done
+          return 1
+        }
+
+        main() {
+          local deploy_name=$(params.SERVICE_NAME)-$(params.NEW_VERSION)
+          local service=$(params.SERVICE_NAME)
+          local version=$(params.NEW_VERSION)
+
+          kubectl apply -f $(params.MANIFEST)
+
+          # Wait until the Deployment is ready by checking the
+          # MinimumReplicasAvailable condition.
+          # Also checking for the ProgressDeadlineExceeded
+          # condition
+          while true; do
+            getMinimumReplicaStatus ${deploy_name}
+            [[ $? -eq 0 ]] && break
+            checkProgressDeadline ${deploy_name}
+            [[ $? -eq 0 ]] && break
+            sleep 5
+          done
+          # If MinimumReplicas are available then check
+          # for the unavailable Pods
+          # Also checking for the ProgressDeadlineExceeded condition
+          while true; do
```

Comment:
can we skip this loop if `checkProgressDeadline` returned true in the first loop

Reply:
effectively setting a flag or 
doing this check
```checkProgressDeadline ${deploy_name}
          if [[ $? -eq 0 ]]; then
            echo "Failed to change the service"
``` 
after the first loop as well